overlord/ifacestate/apparmorprompting/common/common.go

```diff
@@ -336,18 +336,337 @@ func abstractPermissionsToAppArmorFilePermissions(iface string, permissions []st
 	return filePerms, nil
 }
 
-var allowablePathPatternRegexp = regexp.MustCompile(`^(/|(/[^/*{}]+)*(/\*|(/\*\*)?(/\*\.[^/*{}]+)?)?)$`)
+var (
+	// The following must be escaped as literals in a path pattern:
+	problematicChars = `\[\]`
+	// A single safe or escaped unsafe char in a path
+	safePathChar = fmt.Sprintf(`([^%s]|\\[%s])`, problematicChars, problematicChars)
+
+	// The following matches valid path patterns
+	allowablePathPattern       = fmt.Sprintf(`^/%s*$`, safePathChar)
+	allowablePathPatternRegexp = regexp.MustCompile(allowablePathPattern)
+)
+
+// Expands all groups in the given path pattern. Groups are enclosed by '{' '}'.
+// Returns a list of all the expanded path patterns, or an error if the given
+// pattern is invalid.
+func ExpandPathPattern(pattern string) ([]string, error) {
+	patternLen := len(pattern)
```

Comment:
In general, it's okay not to worry about caching `patternLen` as this is just a single go-assembly instruction to read a field. This would also avoid the need to avoid passing it around next to `pattern` :P

Reply:
Good to know, that's much nicer, thanks!